Core/PersistentPixel.swift

```diff
@@ -218,46 +217,49 @@ public final class PersistentPixel: PersistentPixelFiring {
             let queuedPixels = try persistentPixelStorage.storedPixels()
 
             if queuedPixels.isEmpty {
-                self.stopProcessingQueueAndPersistPendingPixels()
+                self.stopProcessingQueue()
                 completion(nil)
                 return
             }
 
             Logger.general.debug("Persistent pixel retrying \(queuedPixels.count, privacy: .public) pixels")
 
-            fire(queuedPixels: queuedPixels) { failedPixels in
+            fire(queuedPixels: queuedPixels) { pixelIDsToRemove in
                 Logger.general.debug("Persistent pixel retrying done, \(queuedPixels.count, privacy: .public) pixels failed and will retry later")
 
                 do {
-                    try self.persistentPixelStorage.replaceStoredPixels(with: failedPixels)
-                    self.stopProcessingQueueAndPersistPendingPixels()
+                    try self.persistentPixelStorage.remove(pixelsWithIDs: pixelIDsToRemove)
+                    self.stopProcessingQueue()
                     completion(nil)
                 } catch {
-                    self.stopProcessingQueueAndPersistPendingPixels()
+                    self.stopProcessingQueue()
                     completion(PersistentPixelStorageError.writeError(error))
                 }
             }
         } catch {
-            self.stopProcessingQueueAndPersistPendingPixels()
+            self.stopProcessingQueue()
             completion(PersistentPixelStorageError.readError(error))
         }
     }
 
     // MARK: - Private
 
-    /// Sends queued pixels and calls the completion handler with any that failed.
-    private func fire(queuedPixels: [PersistentPixelMetadata], completion: @escaping ([PersistentPixelMetadata]) -> Void) {
+    /// Sends queued pixels and calls the completion handler with those that should be removed.
+    private func fire(queuedPixels: [PersistentPixelMetadata], completion: @escaping (Set<UUID>) -> Void) {
         let dispatchGroup = DispatchGroup()
 
-        let failedPixelsAccessQueue = DispatchQueue(label: "Failed Pixel Retry Attempt Metadata Queue")
-        var failedPixels: [PersistentPixelMetadata] = []
+        let pixelIDsAccessQueue = DispatchQueue(label: "Failed Pixel Retry Attempt Metadata Queue")
+        var pixelIDsToRemove: Set<UUID> = []
         let currentDate = dateGenerator()
 
         for pixelMetadata in queuedPixels {
             if let originalSendDateString = pixelMetadata.timestamp,
                let originalSendDate = dateFormatter.date(from: originalSendDateString),
                let date28DaysAgo = calendar.date(byAdding: .day, value: -28, to: currentDate) {
                 if originalSendDate < date28DaysAgo {
+                    pixelIDsAccessQueue.sync {
+                        _ = pixelIDsToRemove.insert(pixelMetadata.id)
+                    }
```

Comment:
We were getting outdated pixels removed for free because they weren't getting added to the failed pixels array, so now we need to explicitly return them as part of the `pixelIDsToRemove`.